### PR TITLE
[CHECKOUT-6901] upgrade typescript version to 4.7

### PR DESCRIPTION
**Pull request by DeveloperAlan**
## What?
Upgrade typescript from 4.1 to 4.7

## Why?
part of our monorepo work, it is required for us to use #971 

## Testing / Proof
N/A

@bigcommerce/checkout

### CODE REVIEW

File: packages/core/src/app/payment/PaymentForm.spec.tsx [hunk 1]
```diff
@@ -247,12 +248,14 @@ describe('PaymentForm', () => {
...
         const handleSubmit = jest.fn();
         const container = mount(<PaymentFormTest
-            { ...defaultProps }
-            onSubmit={ handleSubmit }
-            validationSchema={ getCreditCardValidationSchema({
+            {...defaultProps}
+            onSubmit={handleSubmit}
+            // TODO: need to fix type for this schema
+            // @ts-ignore 
+            validationSchema={getCreditCardValidationSchema({
```
Comment (DeveloperAlan) [inline packages/core/src/app/payment/PaymentForm.spec.tsx:255]:
```
Type 'ObjectSchema<CreditCardFieldsetValues>' is not assignable to type 'ObjectSchema<Partial<PaymentFormValues>>'.
```

Validation schema doesn't match, I was curious what we should do in this situation

**Comment by DeveloperAlan** [comment]
Realised I may have had prettier auto format during save, I'll remove it from this PR

File: packages/core/src/app/payment/PaymentForm.spec.tsx (hunk 1 above)
Comment (DeveloperAlan) [reply packages/core/src/app/payment/PaymentForm.spec.tsx:255]: Will use `as` keyword to fix this

**Comment by DeveloperAlan** [comment]
Currently the eslint in the other PR is blocking the CI/CD to pass

File: .eslintrc.json [hunk 2]
```diff
@@ -41,14 +59,24 @@
...
+        "*.jsx"
+      ],
+      "extends": [
+        "plugin:@nrwl/nx/javascript"
+      ],
       "rules": {}
     }
   ]
-}
+}
```
Comment (animesh1987) [inline .eslintrc.json:82]: EOL

File: packages/core/src/app/checkout/Checkout.tsx [hunk 3]
```diff
@@ -600,6 +604,8 @@ class Checkout extends Component<CheckoutProps & WithCheckoutProps & WithLanguag
             this.setState({ isCartEmpty: true });
 
             if (!isEmbedded()) {
+                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                // @ts-ignore
                 return window.top.location.assign(loginUrl);
```
Comment (animesh1987) [inline packages/core/src/app/checkout/Checkout.tsx:609]: Same as above

File: packages/core/src/app/formFields/getCustomFormFieldsValidationSchema.ts [hunk 4]
```diff
@@ -29,7 +29,7 @@ export interface CustomFormFields {
 
 export default memoize(function getCustomFormFieldsValidationSchema({
     formFields,
-    translate = () => undefined,
+    translate = () => '',
```
Comment (animesh1987) [inline packages/core/src/app/formFields/getCustomFormFieldsValidationSchema.ts:32]: Think we looked at this and agreed to keep it as `undefined` ?

File: packages/core/src/app/payment/Payment.tsx [hunk 5]
```diff
@@ -426,11 +429,12 @@ class Payment extends Component<PaymentProps & WithCheckoutPaymentProps & WithLa
             const order = state.data.getOrder();
             onSubmit(order?.orderId);
         } catch (error) {
-            if (error.type === 'payment_method_invalid') {
+            const handleSubmitError = error as Error & { type: string };
```
Comment (animesh1987) [inline packages/core/src/app/payment/Payment.tsx:432]: Avoid as 😄

File: packages/core/src/app/payment/hostedCreditCard/getHostedCreditCardValidationSchema.spec.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import { HostedCreditCardFieldsetValues } from '@bigcommerce/checkout/payment-integration-api';
+import HostedCreditCardFieldsetValues from '../paymentMethod/HostedCreditCardFieldsetValues'
```
Comment (animesh1987) [inline packages/core/src/app/payment/hostedCreditCard/getHostedCreditCardValidationSchema.spec.ts:1]: Is this intentional?

File: packages/core/src/app/formFields/getCustomFormFieldsValidationSchema.ts (hunk 4 above)
Comment (DeveloperAlan) [reply packages/core/src/app/formFields/getCustomFormFieldsValidationSchema.ts:32]: yep, will change it back to undefined

File: .eslintrc.json (hunk 2 above)
Comment (DeveloperAlan) [reply .eslintrc.json:82]: added in

File: packages/core/src/app/checkout/Checkout.tsx [hunk 7]
```diff
@@ -585,6 +587,8 @@ class Checkout extends Component<CheckoutProps & WithCheckoutProps & WithLanguag
         const { loginUrl, cartUrl, isPriceHiddenFromGuests, isGuestEnabled } = this.props;
 
         if (isPriceHiddenFromGuests) {
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
             return window.top.location.href = cartUrl;
```
[the remark this answers is not in the document]
Comment (DeveloperAlan) [reply packages/core/src/app/checkout/Checkout.tsx:592]: Initially it was to remove the option to a falsey `top` value, but we can just add that in

File: packages/core/src/app/checkout/Checkout.tsx (hunk 3 above)
Comment (DeveloperAlan) [reply packages/core/src/app/checkout/Checkout.tsx:609]: Added as optional property

File: packages/core/src/app/payment/creditCard/getCreditCardValidationSchema.spec.ts [hunk 8]
```diff
@@ -1,4 +1,4 @@
-import { CreditCardFieldsetValues } from '@bigcommerce/checkout/payment-integration-api';
+import CreditCardFieldsetValues from '../../payment/paymentMethod/CreditCardFieldsetValues';
```
[the remark this answers is not in the document]
Comment (DeveloperAlan) [reply packages/core/src/app/payment/creditCard/getCreditCardValidationSchema.spec.ts:1]: Yes it is intentional, as we don't have the `CreditCardFieldsetValues` in that package

File: packages/core/src/app/payment/hostedCreditCard/getHostedCreditCardValidationSchema.spec.ts (hunk 6 above)
Comment (DeveloperAlan) [reply packages/core/src/app/payment/hostedCreditCard/getHostedCreditCardValidationSchema.spec.ts:1]: Yes it is intentional, as we don't have the `HostedCreditCardFieldsetValues` in that package

File: packages/core/src/app/payment/Payment.tsx (hunk 5 above)
Comment (DeveloperAlan) [reply packages/core/src/app/payment/Payment.tsx]: Used a type guard predicate to get over this

File: packages/core/src/app/customer/CheckoutButtonList.tsx [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import { CustomerInitializeOptions, CustomerRequestOptions } from '@bigcommerce/checkout-sdk';
 import React, { memo, Fragment, FunctionComponent } from 'react';
 
+// eslint-disable-next-line import/no-internal-modules
 import { isApplePayWindow } from '../common/utility';
```
[the remark this answers is not in the document]
Comment (DeveloperAlan) [reply packages/core/src/app/customer/CheckoutButtonList.tsx]: Remove the eslint disable and it seemed to work

File: packages/core/src/app/checkout/CheckoutStep.spec.tsx [hunk 10]
```diff
@@ -3,6 +3,7 @@ import { noop } from 'lodash';
 import React from 'react';
 import { CSSTransition } from 'react-transition-group';
 
+// eslint-disable-next-line import/no-internal-modules
 import { isMobileView, MOBILE_MAX_WIDTH } from '../ui/responsive';
```
[the remark this answers is not in the document]
Comment (DeveloperAlan) [reply packages/core/src/app/checkout/CheckoutStep.spec.tsx]: Remove the eslint disable and it seemed to work

File: packages/apple-pay-integration/src/ApplePayPaymentMethod.tsx [hunk 11]
```diff
@@ -14,7 +14,9 @@ const ApplePaymentMethod: FunctionComponent<PaymentMethodProps> = ({ method, che
                     },
                 });
             } catch (error) {
-                onUnhandledError(error);
+                if (error instanceof Error) {
+                    onUnhandledError(error);
+                }
```
Comment (bc-peng) [inline packages/apple-pay-integration/src/ApplePayPaymentMethod.tsx:19]: Just courious, what would happen if there is no `if` block?

File: packages/core/src/app/checkout/CheckoutStep.spec.tsx [hunk 12]
```diff
@@ -46,7 +46,11 @@ describe('CheckoutStep', () => {
     });
 
     afterEach(() => {
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        //@ts-ignore
         delete window.scrollTo;
```
Comment (bc-peng) [inline packages/core/src/app/checkout/CheckoutStep.spec.tsx:51]: It appears a lot of `@ts-ignore` around `window` object. What would be the cost of avoiding `@ts-ignore`?

File: packages/payment-integration-test-framework/src/fixture/pageObject/Checkout.ts [hunk 13]
```diff
@@ -16,7 +16,7 @@ export class Checkout {
     }
 
     // CheckoutFixtures helper, only used in ../CheckoutFixtures.ts
-    async setHarFolderPath(file:string): void {
+    async setHarFolderPath(file:string): Promise<void>  {
```
Comment (bc-peng) [inline packages/payment-integration-test-framework/src/fixture/pageObject/Checkout.ts:19]: This is my mistake. This func is not `async`.
I have removed `async` in today's PR.

File: tests/sampleTests/template.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+        // Playwright actions
+        await checkout.goto();
+        // await page.locator('text=Test Payment ProviderVisaAmexMaster').click();
+        // await page.frameLocator('#bigpaypay-ccNumber iframe').locator('[aria-label="Credit Card Number"]').click();
+        await checkout.placeOrder();
+
+        // Assertions
+        await assertions.shouldSeeOrderConfirmation();
+    });
+});
```
Comment (bc-peng) [inline tests/sampleTests/template.ts:22]: `tests` folder should be already removed.

**Pull request closed without merging**